llvm/lib/Passes/PassBuilderPipelines.cpp

```diff
@@ -968,8 +969,8 @@ PassBuilder::buildInlinerPipeline(OptimizationLevel Level,
   // it's been modified since.
   MainCGPipeline.addPass(createCGSCCToFunctionPassAdaptor(
       RequireAnalysisPass<ShouldNotRunFunctionPassesAnalysis, Function>()));
-
   MainCGPipeline.addPass(CoroSplitPass(Level != OptimizationLevel::O0));
+  MainCGPipeline.addPass(CoroAnnotationElidePass());
```

Comment:
Should this optimization also work for ThinLTO? (also see discussion in https://github.com/llvm/llvm-project/pull/99283#discussion_r1684973257)

I read this pass setup a bit more carefully now, and I don't think it will work for ThinLTO.

Why?

* Both `CoroAnnotationElide` and `CoroSplit` are added to `buildInlinerPipeline`.
* `buildInlinerPipeline` is used by `buildModuleSimplificationPipeline`.
* `buildModuleSimplificationPipeline` is part of both `buildThinLTOPreLinkDefaultPipeline` and `buildThinLTODefaultPipeline`
* -> `CoroSplit` already runs as part of `buildThinLTOPreLinkDefaultPipeline` (i.e. on a per-translation-unit-level)
* -> by the time we reach `buildThinLTODefaultPipeline` (i.e. the cross-TU part of ThinLTO), the coroutines are already split
* -> although, `CoroAnnotationElide` is run a 2nd time as part of cross-TU optimization, it will be a no-op due to the `Caller->isPresplitCoroutine()` check in `CoroAnnotationElide.cpp:120`

Reply:
There is another PR #90310 trying to move the coro passes into post-link pipeline if ThinLTO(`-flto=thin`) is enabled for the compilation. It ran into some issue with asan and going through some refactoring.

Reply:
Thanks for that context! I assume, as part of relanding #90310 you would then also adjust the way the `CoroAnnotationElide` registration?

Do you happen to know the answer to https://github.com/llvm/llvm-project/pull/99283#discussion_r1684973257 or can provide some guidance there?

Reply:
Yes, the pipeline will need to be adjusted for the new pass.